File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
             BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out))) {
            String[] input = br.readLine().split(" ");
            int N = Integer.parseInt(input[0]), C = Integer.parseInt(input[1]);
            List<Integer> houses = new ArrayList<>();
            for (int i = 0; i < N; i++) {
                houses.add(Integer.parseInt(br.readLine()));
            }
            Collections.sort(houses);
            int low = 1, high = houses.get(N - 1) - houses.get(0) + 1;
            while (low < high) {
                int mid = (low + high) >>> 1;
                if (C <= countHouses(houses, mid)) low = mid + 1;
                else high = mid;
            }
            bw.write(String.valueOf(low - 1));
            bw.flush();
        }
    }
    private static int countHouses(List<Integer> houses, int distance) {
        int pointer = 0, answer = 1, n = houses.size();
        for (int i = 0; i < n; i++) {
            if (houses.get(i) - houses.get(pointer) >= distance) {
                pointer = i;
                answer++;
            }
        }
        return answer;
    }
}

File: 백준/Gold/2110. 공유기 설치/공유기 설치.cc
#include <iostream>
#include <vector>
#include <algorithm>

#define FAST_IO cin.tie(NULL)->ios::sync_with_stdio(false);

using namespace std;

#define REP(i, a, b) for (int i = a; i <= b; i++)

int count_houses(const vector<int> &houses, int distance) {
    int pointer = 0, answer = 1, n = houses.size();

    REP(i, 0, n - 1) {
        if(houses[i] - houses[pointer] >= distance) {
            pointer = i;
            answer++;
        }
    }
    return answer;
}

int main() {
    FAST_IO

    int n, c; cin >> n >> c;
    vector<int> houses(n);
    REP(i, 0, n - 1) cin >> houses[i];
    sort(houses.begin(), houses.end());
    int low = 1,
        high = houses[n - 1] - houses[0],
        answer = 0;

    while(low <= high) {
        int mid = (low + high) / 2;
        if(count_houses(houses, mid) >= c) {
            answer = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    cout << answer;
    return EXIT_SUCCESS;
}